curvefs/src/metaserver/copyset/copyset_node.cpp

```diff
@@ -0,0 +1,509 @@
+                          iter.index(), doneGuard.release(),
+                          TimeUtility::GetTimeofDayUs());
+            applyQueue_->Push(metaClosure->GetOperator()->HashCode(),
+                              std::move(task));
+            LOG(INFO) << "on_apply, term " << iter.term();
+        } else {
+            // 从log中解析出具体的请求
+            auto metaOperator = RaftLogCodec::Decode(this, iter.data());
+            CHECK(metaOperator != nullptr) << "Decode raft log failed";
+            auto hashcode = metaOperator->HashCode();
+            auto task =
+                std::bind(&MetaOperator::OnApplyFromLog, metaOperator.release(),
+                          TimeUtility::GetTimeofDayUs());
+            applyQueue_->Push(hashcode, std::move(task));
+            LOG(INFO) << "on_apply from log, term " << iter.term();
+        }
+    }
+}
+
+void CopysetNode::on_shutdown() {
+    LOG(INFO) << "Copyset: " << name_ << " is shutdown";
+}
+
+class OnSnapshotSaveDoneImpl : public OnSnapshotSaveDone {
+ public:
+    OnSnapshotSaveDoneImpl(CopysetNode* node, braft::SnapshotWriter* writer,
+                           braft::Closure* snapDone, uint64_t startTimeUs)
+        : node_(node),
+          writer_(writer),
+          snapDone_(snapDone),
+          statusCode_(MetaStatusCode::OK),
+          startTimeUs_(startTimeUs) {}
+
+    void Run() override {
+        std::unique_ptr<OnSnapshotSaveDoneImpl> selfGuard(this);
+        brpc::ClosureGuard doneGuard(snapDone_);
+
+        if (statusCode_ != MetaStatusCode::OK) {
+            RaftSnapshotMetric::GetInstance().OnSnapshotSaveFailure();
+            return;
+        }
+
+        writer_->add_file(kMetaDataFilename);
+        RaftSnapshotMetric::GetInstance().OnSnapshotSaveComplete(
+            TimeUtility::GetTimeofDayUs() - startTimeUs_);
+    }
+
+    void SetSuccess() override {
+        LOG(INFO) << "Copyset " << node_->Name() << " save snapshot success";
+    }
+
+    void SetError(MetaStatusCode code) override {
+        statusCode_ = code;
+        snapDone_->status().set_error(statusCode_,
+                                      "Save metadata failed, err: %s",
+                                      MetaStatusCode_Name(statusCode_).c_str());
+        LOG(ERROR) << "Copyset " << node_->Name()
+                   << " save snapshot failed, err: "
+                   << MetaStatusCode_Name(statusCode_);
+    }
+
+ private:
+    CopysetNode* node_;
+    braft::SnapshotWriter* writer_;
+    braft::Closure* snapDone_;
+    MetaStatusCode statusCode_;
+    uint64_t startTimeUs_;
+};
+
+void CopysetNode::on_snapshot_save(braft::SnapshotWriter* writer,
+                                   braft::Closure* done) {
+    LOG(INFO) << "Copyset " << name_ << " saving snapshot to '"
+              << writer->get_path() << "'";
+
+    brpc::ClosureGuard doneGuard(done);
+    auto startTimeUs = TimeUtility::GetTimeofDayUs();
+
+    // flush all flying operators
+    applyQueue_->Flush();
```

Comment:
because we use apply queue to do apply tasks, so when raft triggers snapshot, we should make sure that tasks in apply queue had been done which means metastore has complete metadata before current snapshot point.

Reply:
ok